### DISCUSSION TITLE
How to inject go struct into cue build instance?

### DISCUSSION THREAD

**Discussion by joscha-alisch** [cuelang/cue#710] [Q&A] [answered]
Say, I have a cue file like this
```yaml
test: fromGo.value
```

I want to inject a struct from go into this file:
```go
struct FromGo {
   Value: string
}
...
fromGo := FromGo{Value: "something"}
``` 

So that the the cue file evaluates to
```yaml
test: "something"
```

If I create another cue file that acts a a schema for the struct, load both files and `use instance.Fill`, it works fine, like so 
```go
b := &build.Instance{}

err := b.AddFile("test.cue", "test: fromGo.value")
panicErr(err)

err = b.AddFile("schema.cue", "fromGo: { value: string }")
panicErr(err)

r := cue.Runtime{}
i, err := r.Build(b)
panicErr(err)

fromGo := FromGo{Value: "something"}
i, err = i.Fill(&fromGo, "fromGo")
panicErr(err)

s, err := i.Lookup("test").String()
panicErr(err)
println(s) // prints "something"

```

But I wonder whether there is a way to skip having to handcraft this second schema file? I.e. replace `b.AddFile("schema.cue", "fromGo: { value: string }")` with something automatically generated from the struct type. Given that there is the gocodec package I assume there is some way?

**Comment by mpvl** [comment, marked as answer]
The API could be improved. But from gocodec you can get a cue.Value, which has a Syntax method to get an AST, from which you can get a File using astutil.ToFile.

**Reply by joscha-alisch** [reply]
Ah perfect, I got to the point where I found the Syntax method but wasn't able to find out how to make a file out of it. 
Thanks @mpvl !

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/710.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
